src/printer.js

```diff
@@ -78,12 +78,26 @@ function genericPrint(path, options, printPath, args) {
     // responsible for printing node.decorators.
     !util.getParentExportDeclaration(path)
   ) {
-    const separator = node.decorators.length === 1 &&
-      (node.decorators[0].expression.type === "Identifier" ||
-        node.decorators[0].expression.type === "MemberExpression")
-      ? " "
-      : hardline;
+    let shouldBeInlined = false;
+    if (node.decorators.length === 1) {
+      let decorator = node.decorators[0];
+      if (node.decorators[0].expression) {
+        decorator = node.decorators[0].expression;
+      }
+      if (
+        decorator.type === "Identifier" ||
+        decorator.type === "MemberExpression"
+      ) {
+        shouldBeInlined = true;
+      }
+    }
+
+    const separator = shouldBeInlined ? " " : hardline;
     path.each(function(decoratorPath) {
+      const value = decoratorPath.getValue();
+      if (value.type !== "Decorator" && value.type !== "TSDecorator") {
```

Comment:
For
```js
@dec enum E {}
```

it is wrapped in a TSDecorator. So the wrapping is inconsistent. Also, it would be nice to name it Decorator and drop the TS prefix.